Mathlib/ModelTheory/ElementarySubstructures.lean

```diff
@@ -0,0 +1,175 @@
+
+/-- A substructure is elementary when every formula applied to a tuple in the subtructure
+  agrees with its value in the overall structure. -/
+def IsElementary (S : L.Substructure M) : Prop :=
+  ∀ ⦃n⦄ (φ : L.Formula (Fin n)) (x : Fin n → S), φ.Realize (((↑) : _ → M) ∘ x) ↔ φ.Realize x
+#align first_order.language.substructure.is_elementary FirstOrder.Language.Substructure.IsElementary
+
+end Substructure
+
+/-- An elementary substructure is one in which every formula applied to a tuple in the subtructure
+  agrees with its value in the overall structure. -/
+structure ElementarySubstructure where
+  toSubstructure : L.Substructure M
+  isElementary' : toSubstructure.IsElementary
+#align first_order.language.elementary_substructure FirstOrder.Language.ElementarySubstructure
+#align first_order.language.elementary_substructure.to_substructure FirstOrder.Language.ElementarySubstructure.toSubstructure
+#align first_order.language.elementary_substructure.is_elementary' FirstOrder.Language.ElementarySubstructure.isElementary'
+
+variable {L M N}
+
+namespace ElementarySubstructure
+
+attribute [coe] toSubstructure
+
+instance instCoe : Coe (L.ElementarySubstructure M) (L.Substructure M) :=
+  ⟨ElementarySubstructure.toSubstructure⟩
+#align first_order.language.elementary_substructure.first_order.language.substructure.has_coe FirstOrder.Language.ElementarySubstructure.instCoe
+
+instance instSetLike : SetLike (L.ElementarySubstructure M) M :=
+  ⟨fun x => x.toSubstructure.carrier, fun ⟨⟨s, hs1⟩, hs2⟩ ⟨⟨t, ht1⟩, _⟩ _ => by
+    congr⟩
+#align first_order.language.elementary_substructure.set_like FirstOrder.Language.ElementarySubstructure.instSetLike
+
+instance inducedStructure (S : L.ElementarySubstructure M) : L.Structure S :=
+  Substructure.inducedStructure
+set_option linter.uppercaseLean3 false in
+#align first_order.language.elementary_substructure.induced_Structure FirstOrder.Language.ElementarySubstructure.inducedStructure
+
+@[simp]
+theorem isElementary (S : L.ElementarySubstructure M) : (S : L.Substructure M).IsElementary :=
+  S.isElementary'
+#align first_order.language.elementary_substructure.is_elementary FirstOrder.Language.ElementarySubstructure.isElementary
+
+/-- The natural embedding of an `L.Substructure` of `M` into `M`. -/
+def subtype (S : L.ElementarySubstructure M) : S ↪ₑ[L] M where
+  toFun := (↑)
+  map_formula' := S.isElementary
+#align first_order.language.elementary_substructure.subtype FirstOrder.Language.ElementarySubstructure.subtype
+
+@[simp]
+theorem coeSubtype {S : L.ElementarySubstructure M} : ⇑S.subtype = ((↑) : S → M) :=
+  rfl
+#align first_order.language.elementary_substructure.coe_subtype FirstOrder.Language.ElementarySubstructure.coeSubtype
+
+/-- The substructure `M` of the structure `M` is elementary. -/
+instance instTop : Top (L.ElementarySubstructure M) :=
+  ⟨⟨⊤, fun _ _ _ => (Substructure.realize_formula_top _ _).symm⟩⟩
```

Comment:
This proof grew two underscores; was that deliberate, or should the arguments to `realize_formula_top` have been kept implicit?

If it was deliberate, please put it in the PR description.

bors d+

Reply:
It was not deliberate - I have made those variables implicit again.
In fact, I have now moved those two lemmas to the Substructures file, as I think they really belong there. I will adjust the PR description accordingly.